src/common/policy.c

```diff
@@ -0,0 +1,154 @@
+/**
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * This file is part of daos. It implements some miscellaneous policy functions
+ */
+
+#include <daos_srv/policy.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#define POLICY_BAD_PARAM	(-1)
+
+static const char *POLICY_NAMES[DAOS_MEDIA_POLICY_MAX] = {
+	"io_size",
+	"write_intensivity" };
+
+static const char *POLICY_PARAMS[DAOS_MEDIA_POLICY_MAX]
+				[DAOS_MEDIA_POLICY_PARAMS_MAX] = {
+{"th1", "th2", "", ""},				/* io_size */
+{"wr_size", "hot1", "hot2", ""}			/* write_intensivity */
+};
+
+static const char *PARAM_DELIM = ",";
+static const char *VALUE_DELIM = "=";
+
+static bool
+parse_param_val(int param_idx, char *tok, struct policy_desc_t *out_pd)
+{
+	char *endptr;
+	int val = strtol(tok, &endptr, 10);
+
+	if (*tok != '\0' && *endptr == '\0') {
+		if (out_pd != NULL)
+			out_pd->params[param_idx] = val;
+		return true;
+	}
+
+	return false;
+}
+
+static int
+get_param_idx_for_policy(int policy_idx, char *name, int len)
+{
+	int param_idx = POLICY_BAD_PARAM;
+	int i;
+
+	for (i = 0; i < DAOS_MEDIA_POLICY_PARAMS_MAX; i++) {
+		if (strncmp(name, POLICY_PARAMS[policy_idx][i], len) == 0) {
+			param_idx = i;
+			break;
+		}
+	}
+
+	return param_idx;
+}
+
+static int
+policy_name_to_idx(char *name, int len)
+{
+	int policy_idx = DAOS_MEDIA_POLICY_MAX;
+	int i;
+
+	for (i = 0; i < DAOS_MEDIA_POLICY_MAX; i++) {
+		if (strncmp(name, POLICY_NAMES[i], len) == 0) {
+			policy_idx = i;
+			break;
+		}
+	}
+
+	return policy_idx;
+}
+
+bool
+daos_policy_try_parse(const char *policy_str, struct policy_desc_t *out_pd)
+{
+	char		*tok = NULL;
+	const char	*delim = PARAM_DELIM;
+	const char	*delim2 = VALUE_DELIM;
+	char		*saveptr = NULL;
+	char		*saveptr2 = NULL;
+	char		*str2 = NULL;
+	bool		type_found = false;
+	int		param_idx = POLICY_BAD_PARAM;
+	int		tok_len = 0;
+	bool		ret_val = false;
+	int		i, j;
+	size_t		len;
+	char		*str;
+	unsigned int	policy_idx = DAOS_MEDIA_POLICY_MAX;
+
+	len = strlen(policy_str);
+	D_STRNDUP(str, policy_str, len); /* for strtok_r */
+
+	/* tokenize input to "name=value" format */
+	for (i = 0; ; i++, str = NULL) {
```

Comment:
'str' is set to NULL here, I think that's why NLT warns memory leaked on 'str'. why do we have to dup the 'policy_str' instead of using it directly?

Reply:
We have to duplicate it because "strtok_r" modifies it's arguments (e.g. inserts '\0' for the delimiter), I will store the duplicated string pointer just to free it after tokenizing.